src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,231 @@
+
+package index
+
+import (
+	"errors"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// ErrWideQueryResultsExhausted is used to short circuit additional document
+// entries being added if theese wide results will no longer accept documents,
+// e.g. if the results are closed, or if no further docuemnts will pass the
+// shard filter.
+var ErrWideQueryResultsExhausted = errors.New("no more values to add to wide query results")
+
+type shardFilterFn func(ident.ID) (uint32, bool)
+
+type wideResults struct {
+	nsID   ident.ID
+	idPool ident.Pool
+
+	closed      bool
+	idsOverflow []ident.ID
+	batch       *ident.IDBatch
+	batchCh     chan<- *ident.IDBatch
+	batchSize   int
+
+	shardFilter shardFilterFn
+	shards      []uint32
+	shardIdx    int
+	// NB: pastLastShard will mark this reader as exhausted after a
+	// document is discovered whose shard exceeds the last shard this results
+	// is responsible for, using the fact that incoming documents are sorted by
+	// shard then by ID.
+	pastLastShard bool
+}
+
+// NewWideQueryResults returns a new wide query results object.
+// NB: Reader must read results from `batchCh` in a goroutine, and call
+// batch.Done() after the result is used, and the writer must close the
+// channel after no more Documents are available.
+func NewWideQueryResults(
+	namespaceID ident.ID,
+	idPool ident.Pool,
+	shardFilter shardFilterFn,
+	opts WideQueryOptions,
+) BaseResults {
+	batchSize := opts.BatchSize
+	results := &wideResults{
+		nsID:        namespaceID,
+		idPool:      idPool,
+		batchSize:   batchSize,
+		idsOverflow: make([]ident.ID, 0, batchSize),
+		batch: &ident.IDBatch{
+			IDs: make([]ident.ID, 0, batchSize),
+		},
+		batchCh: opts.IndexBatchCollector,
+		shards:  opts.ShardsQueried,
+	}
+
+	if len(opts.ShardsQueried) > 0 {
+		// Only apply filter if there are shards to filter against.
+		results.shardFilter = shardFilter
+	}
+
+	return results
+}
+
+func (r *wideResults) AddDocuments(batch []doc.Document) (int, int, error) {
+	if r.closed || r.pastLastShard {
+		return 0, 0, ErrWideQueryResultsExhausted
+	}
+
+	err := r.addDocumentsBatchWithLock(batch)
+	release := len(r.batch.IDs) == r.batchSize
+	if release {
+		r.releaseAndWait()
+		r.releaseOverflow(false)
+	}
+
+	return 0, 0, err
+}
+
+func (r *wideResults) releaseOverflow(forceRelease bool) {
+	var (
+		incomplete bool
+		size       int
+		overflow   int
+	)
+	for {
+		size = r.batchSize
+		overflow = len(r.idsOverflow)
+		if overflow == 0 {
+			// NB: no overflow elements.
+			return
+		}
+
+		if overflow < size {
+			size = overflow
+			incomplete = true
+		}
+
+		r.batch.IDs = append(r.batch.IDs, r.idsOverflow[0:size]...)
+		r.batch.IDs = r.batch.IDs[:size]
+		copy(r.idsOverflow, r.idsOverflow[size:])
+		r.idsOverflow = r.idsOverflow[:overflow-size]
+		if !forceRelease && incomplete {
+			return
+		}
+
+		r.releaseAndWait()
+	}
+}
+
+func (r *wideResults) addDocumentsBatchWithLock(batch []doc.Document) error {
```

Comment:
What's the protocol for these `...WithLock` methods? I'd expect them to either eventually lock at some point or to be called while locked, but it doesn't seem that either case is true as these methods bubble up to `AddDocument` which doesn't do any locking.

Reply:
It's the latter, they should be called within a lock, good catch!